### PR TITLE
Alias mapping for group by operator. 

### PR DESCRIPTION
**Pull request by IshSiva**

### CODE REVIEW

File: eva/models/storage/batch.py [hunk 1]
```diff
@@ -244,15 +244,16 @@ def project(self, cols: None) -> Batch:
...
+        
+        # verfied_cols = [c for c in cols if c in self._frames]
+        # unknown_cols = list(set(cols) - set(verfied_cols))
+        # if len(unknown_cols):
+        #     logger.warn(
+        #         "Unexpected columns %s\n\
+        #                          Frames: %s"
+        #         % (unknown_cols, self._frames)
+        #     )
+        return Batch(self._frames.iloc[:, cols], self._identifier_column)
```
Comment (gaurav274) [inline eva/models/storage/batch.py:256]: Won't this operation slow down the performance?

File: eva/models/storage/batch.py [hunk 2]
```diff
@@ -446,3 +447,7 @@ def drop_column_alias(self) -> None:
                 new_col_names.append(col_name)
 
         self._frames.columns = new_col_names
+
+    def reorder_columns(self, col_order):
```
Comment (gaurav274) [inline eva/models/storage/batch.py:451]: What is `col_order`? Add the typing

File: eva/utils/single_dispatch_method.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import sys
+
+if sys.version_info >= (3, 8):
+    from functools import singledispatchmethod
+else:
+    # https://stackoverflow.com/questions/24601722/how-can-i-use-functools-singledispatch-with-instance-methods
+    from functools import singledispatch, update_wrapper
+
+    def singledispatchmethod(func):
```
Comment (LordDarkula) [inline eva/utils/single_dispatch_method.py:9]: Add docstring describing what single dispatch is briefly and provide short code example.

File: eva/optimizer/column_deriver.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+class ColumnDeriver:
+    """
+    Derives the id for the column names in the optimal plans
+    """
+
+    def __init__(self):
+        pass
+    
+    @singledispatchmethod
+    def scan_node_for_attributes(self, node:AbstractPlan, alias_prefix:str=None) -> dict:
```
Comment (LordDarkula) [inline eva/optimizer/column_deriver.py:22]: I would write a single function and use `isinstance` checks to decide what to do depending on input types. It is easier to read and more Pythonic.

File: eva/optimizer/column_deriver.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        
+        prev_merged = {}
+        for child in children:
+            curr_map, merged = self.scan_node_for_attributes(child, alias_prefix)
+            prev_map.update(curr_map)
+            prev_merged.update(merged)
+        
+        return prev_map, prev_merged
+
+    def merge(self, map1, map2):
```
Comment (LordDarkula) [inline eva/optimizer/column_deriver.py:40]: Add type annotations.

File: eva/optimizer/column_mapper.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    """
+    Maps the ids for columns
+    """
+    
+    def __init__(self, column_to_id_map: dict, updated_col_map: dict):
+        self.column_to_id_map = column_to_id_map
+        self.updated_col_map = updated_col_map
+
+    @singledispatchmethod
+    def map_node_attributes_to_id(self, node: AbstractPlan):
```
Comment (LordDarkula) [inline eva/optimizer/column_mapper.py:24]: Same comment as above. Use `isinstance` instead.

File: eva/executor/seq_scan_executor.py [hunk 7]
```diff
@@ -44,8 +44,8 @@ def exec(self, **kwargs) -> Iterator[Batch]:
         for batch in child_executor.exec(**kwargs):
             # apply alias to the batch
             # id, data -> myvideo.id, myvideo.data
-            if self.alias:
-                batch.modify_column_alias(self.alias)
+            # if self.alias:
```
Comment (LordDarkula) [inline eva/executor/seq_scan_executor.py:47]: Add note explaining why code is commented out.

File: eva/models/storage/batch.py [hunk 8]
```diff
@@ -244,15 +244,16 @@ def project(self, cols: None) -> Batch:
...
-        unknown_cols = list(set(cols) - set(verfied_cols))
-        if len(unknown_cols):
-            logger.warn(
-                "Unexpected columns %s\n\
-                                 Frames: %s"
-                % (unknown_cols, self._frames)
-            )
-        return Batch(self._frames[verfied_cols])
+        
+        # verfied_cols = [c for c in cols if c in self._frames]
```
Comment (LordDarkula) [inline eva/models/storage/batch.py:248]: Add note explaining why code is commented out.

File: eva/optimizer/column_deriver.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+from eva.utils.single_dispatch_method import singledispatchmethod
+from eva.optimizer.column_mapper import ColumnMapper
+
+
+class ColumnDeriver:
+    """
+    Derives the id for the column names in the optimal plans
+    """
+
+    def __init__(self):
```
Comment (LordDarkula) [inline eva/optimizer/column_deriver.py:18]: Unecessary

File: eva/utils/single_dispatch_method.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import sys
+
+if sys.version_info >= (3, 8):
```
Comment (LordDarkula) [inline eva/utils/single_dispatch_method.py:3]: Add note explaining the reasoning behind the python version check

**Review by LordDarkula — CHANGES_REQUESTED** [review]
The functionality of the code is fine. I would like to see more type hints and documentation explaining the reasoning behind some of the choices made.

File: eva/models/storage/batch.py (hunk 2 above)
Comment (IshSiva) [reply eva/models/storage/batch.py]: col_order is list. Updated.

File: eva/utils/single_dispatch_method.py (hunk 3 above)
Comment (IshSiva) [reply eva/utils/single_dispatch_method.py:10]: done.

File: eva/optimizer/column_deriver.py (hunk 4 above)
Comment (IshSiva) [reply eva/optimizer/column_deriver.py:18]: This design decision was discussed with Gaurave and we decided to use the visitor pattern. Hence the above method was used.

File: eva/executor/seq_scan_executor.py (hunk 7 above)
Comment (IshSiva) [reply eva/executor/seq_scan_executor.py]: the comments have been removed. The code logic has been updated so the commented block is not needed anymore,

File: eva/models/storage/batch.py (hunk 8 above)
Comment (IshSiva) [reply eva/models/storage/batch.py]: Removed the commented block as its not needed.

File: eva/utils/single_dispatch_method.py (hunk 10 above)
Comment (IshSiva) [reply eva/utils/single_dispatch_method.py:4]: done.

File: eva/optimizer/column_deriver.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        #update column numbers
+        
+        attribute_id_counter = len(col_update_map)
+
+        for key,_ in right_child_map.items():
+            if key not in col_update_map:
+                col_update_map[key] = attribute_id_counter
+                attribute_id_counter+=1
+        
+        return merged_id_map, col_update_map
```
Comment (gaurav274) [inline eva/optimizer/column_deriver.py:103]: Is the second output ever used?

File: eva/optimizer/plan_generator.py [hunk 12]
```diff
@@ -84,6 +84,11 @@ def optimize(self, logical_plan: Operator):
 
         # Build Optimal Tree
         optimal_plan = self.build_optimal_physical_plan(root_grp_id, optimizer_context)
+
+        #map columns to ids
```
Comment (gaurav274) [inline eva/optimizer/plan_generator.py:88]: Please elaborate the comment

File: eva/optimizer/column_mapper.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+from eva.expression.tuple_value_expression import TupleValueExpression
+from eva.expression.constant_value_expression import ConstantValueExpression
+from eva.expression.abstract_expression import AbstractExpression
+from eva.planner.abstract_plan import AbstractPlan
+from eva.utils.single_dispatch_method import singledispatchmethod
+
+
+class ColumnMapper:
+    """
+    Maps the ids for columns
```
Comment (gaurav274) [inline eva/optimizer/column_mapper.py:16]: Elaborate to better explain the role of the file

File: eva/optimizer/column_mapper.py (hunk 6 above)
Comment (gaurav274) [inline eva/optimizer/column_mapper.py:24]: This function name is unintuitive. How about `column_names_to_idx` ?

File: eva/optimizer/column_deriver.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+from eva.utils.single_dispatch_method import singledispatchmethod
+from eva.optimizer.column_mapper import ColumnMapper
+
+
+class ColumnDeriver:
+    def __init__(self):
+        pass
+    
+    @singledispatchmethod
+    def scan_node_for_attributes(self, node:AbstractPlan, alias_prefix:str=None) -> dict:
```
Comment (gaurav274) [inline eva/optimizer/column_deriver.py:18]: In the context of DB, the scan keyword makes it confusing. 

Can we rename it to maybe `derive_column_mapping`?

**Review by gaurav274 — CHANGES_REQUESTED** [review]

File: eva/optimizer/column_deriver.py (hunk 11 above)
Comment (IshSiva) [reply eva/optimizer/column_deriver.py:103]: yes its used in the project plan on line 45 in ColumnMapper

File: eva/optimizer/plan_generator.py (hunk 12 above)
Comment (IshSiva) [reply eva/optimizer/plan_generator.py]: updated it. please let me know if its okay

File: eva/optimizer/column_mapper.py (hunk 13 above)
Comment (IshSiva) [reply eva/optimizer/column_mapper.py]: done

File: eva/optimizer/column_deriver.py (hunk 14 above)
Comment (IshSiva) [reply eva/optimizer/column_deriver.py]: done.

**Comment by jarulraj** [comment]
@IshSiva Is this PR still relevant?

**Comment by IshSiva** [comment]
Closing this for now as we have to make many changes. Can be created as a new PR in the future.

**Pull request closed without merging**